submitter/tx_submitter.go

```diff
@@ -99,25 +114,45 @@ func (s *TxSubmitter) submitForSingleEvent(event *model.Event) error {
 	}
 
 	// submit transaction
-	txHash, errTx := s.SubmitTx(event, valBitSet, aggregatedSignature)
-	if errTx != nil {
-		logging.Logger.Errorf("failed to submit tx for challengeId: %d,  err: %s", event.ChallengeId, errTx.Error())
-	} else {
-		logging.Logger.Infof("submitter tx submitted for challengeId: %d, txHash: %s ,timestamp: %s", event.ChallengeId, txHash, time.Now().Format("15:04:05.000000"))
+	// TODO: check TxOption
+	submittedAttempts := 0
+	for {
+		if submittedAttempts > common.MaxSubmitAttempts {
+			return fmt.Errorf("submitter exceeded max submit attempts for challengeId: %d, timestamp: %s, err=%+v", event.ChallengeId, time.Now().Format("15:04:05.000000"), err.Error())
+		}
+		attestRes, err := s.executor.AttestChallenge(s.executor.GetAddr(), event.ChallengerAddress, event.SpOperatorAddress, event.ChallengeId, math.NewUintFromString(event.ObjectId), challengetypes.CHALLENGE_SUCCEED, valBitSet.Bytes(), aggregatedSignature, types.TxOption{})
+		if err != nil {
+			submittedAttempts++
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+		if !attestRes {
+			submittedAttempts++
+			time.Sleep(100 * time.Millisecond)
```

Comment:
If it is not its turn after some sleep? 
`QueryInturnAttestationSubmitter` not only returns who can submit, it also returns the timespan (start, end). You can also utilize `end` here.

Reply:
fixed